sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/Exchange.scala

```diff
@@ -95,46 +89,3 @@ case class ReusedExchangeExec(override val output: Seq[Attribute], child: Exchan
        |""".stripMargin
   }
 }
-
-/**
- * Find out duplicated exchanges in the spark plan, then use the same exchange for all the
- * references.
- */
-case class ReuseExchange(conf: SQLConf) extends Rule[SparkPlan] {
-
-  def apply(plan: SparkPlan): SparkPlan = {
-    if (!conf.exchangeReuseEnabled) {
-      return plan
-    }
-    // Build a hash map using schema of exchanges to avoid O(N*N) sameResult calls.
-    val exchanges = mutable.HashMap[StructType, ArrayBuffer[Exchange]]()
-
-    // Replace a Exchange duplicate with a ReusedExchange
-    def reuse: PartialFunction[Exchange, SparkPlan] = {
-      case exchange: Exchange =>
-        val sameSchema = exchanges.getOrElseUpdate(exchange.schema, ArrayBuffer[Exchange]())
-        val samePlan = sameSchema.find { e =>
-          exchange.sameResult(e)
-        }
-        if (samePlan.isDefined) {
-          // Keep the output of this exchange, the following plans require that to resolve
-          // attributes.
-          ReusedExchangeExec(exchange.output, samePlan.get)
-        } else {
-          sameSchema += exchange
-          exchange
-        }
-    }
-
-    plan transformUp {
-      case exchange: Exchange => reuse(exchange)
-    } transformAllExpressions {
-      // Lookup inside subqueries for duplicate exchanges
-      case in: InSubqueryExec =>
-        val newIn = in.plan.transformUp {
-          case exchange: Exchange => reuse(exchange)
-        }
-        in.copy(plan = newIn.asInstanceOf[BaseSubqueryExec])
-    }
-  }
```

Comment:
This `transformUp` and then `transformAllExpressions` was actually a bit weird for 2 reasons:
- A minor issue is that `reuse` is a partial function and yet we use it as a normal function, I mean this code probably should have been written as 
  ```
    plan transformUp reuse transformAllExpressions {
      // Lookup inside subqueries for duplicate exchanges
      case in: InSubqueryExec =>
        val newIn = in.plan.transformUp reuse
        in.copy(plan = newIn.asInstanceOf[BaseSubqueryExec])
    }
  ```
  and the definition of `reuse` should have been `def reuse: PartialFunction[SparkPlan, SparkPlan] = {`
- A bigger issue is this part of the code traverses the plan 2 times, first with the `transformUp` and then with the `transformAllExpressions` and both traversals can insert reuse references to the plan. Imagine that the first traversal (`transformUp`) inserts a reuse reference to an exchange. Let's say it is inserts a `ReuseExchange` node pointing to the `Exchange id=1` but then the second traversal (`transformAllExpressions`) finds an `InSubqueryExec` expression in a `FileSourceScanExec` under the node `Exchange id=1`. If there is reuse opportunity in the subplan of that `InSubqueryExec` then the change due to inserting another reuse node to the subplan propagates up and results and `Exchange id=1` will be replaced in parent plan and reuse node crated in the first traversal pointing to it will be invalid.
  (Please note that this issue is very similar to what I showed in the description of the PR but in that case the 2 traversals were due to the separate `ReuseSubquery` and `ReuseExchange` rules.)
  I think the fix to this issue is to use an 1 pass, whole-plan, bottom-up traversal like I did in `ReuseExchangeAndSubquery` in this PR.

Reply:
can you tell me the exact test case that demonstrates this issue?

I kinda see the issue you're talking about, but my thinking is that maybe this just requires fixing `ReuseExchange`, rather than combining both into one rule.

Reply:
This is the 1. point in the description and tested by the case `SPARK-32041: No reuse interference inside ReuseExchange` in the new `ReuseExchangeAndSubquerySuite`: https://github.com/apache/spark/pull/28885/files#diff-f6f54d5cfc4254d8ed9122013394351bR28